consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -0,0 +1,199 @@
+	identities, err := c.state.AtBlockID(blockID).Identities(filter.And(
+		consensusMemberFilter,
+		selector,
+	))
+	return identities, err
+}
+
+func (c *Consensus) Identity(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+	identities, err := c.Identities(blockID, filter.HasNodeID(nodeID))
+	if err != nil {
+		return nil, fmt.Errorf("could not get identity (id=%s): %w", nodeID, err)
+	}
+	if len(identities) != 1 {
+		return nil, fmt.Errorf("found invalid number (%d) of identities for node id (%s)", len(identities), nodeID)
+	}
+	return identities[0], nil
+}
+
+func (c *Consensus) LeaderForView(view uint64) (flow.Identifier, error) {
+
+	// STEP 1 - look for an epoch matching this view for which we have already
+	// pre-computed leader selection. Epochs last ~500k views, so we find the
+	// epoch here 99.99% of the time. Since epochs are long-lived, it is fine
+	// for this to be linear in the number of epochs we have observed.
+	for _, epoch := range c.leaders {
+		if view >= epoch.selection.FirstView() && view <= epoch.selection.FinalView() {
+			index, err := epoch.selection.LeaderIndexForView(view)
+			if err != nil {
+				return flow.ZeroID, fmt.Errorf("could not get leader index for view: %w", err)
+			}
+			return epoch.identities[index].NodeID, nil
+		}
+	}
+
+	// STEP 2 - we haven't yet computed leader selection for an epoch containing
+	// the requested view. We compute leader selection for the current and previous
+	// epoch (w.r.t. the finalized head) at initialization then compute leader
+	// selection for the next epoch when we encounter any view for which we don't
+	// know the leader. The series of epochs we have computed leaders for is
+	// strictly consecutive, meaning we know the leader for all views V where:
+	//
+	//   V >= oldestEpoch.firstView && V <= newestEpoch.finalView
+	//
+	// Thus, the requested view is either before oldestEpoch.firstView or after
+	// newestEpoch.finalView.
+	//
+	// CASE 1: V < oldestEpoch.firstView
+	// If the view is before the first view we've computed the leader for, this
+	// represents an invalid query because we only guarantee the protocol state
+	// will contain epoch information for the current, previous, and next epoch -
+	// such a query must be for a view within an epoch at least TWO epochs before
+	// the current epoch when we started up. This is considered an invalid query.
+	//
+	// CASE 2: V > newestEpoch.finalView
+	// If the view is after the last view we've computed the leader for, we
+	// assume the view is within the next epoch (w.r.t. the finalized head).
+	// This assumption is equivalent to assuming that we build at least one
+	// block in every epoch, which is anyway a requirement for valid epochs.
+	//
+	next := c.state.Final().Epochs().Next()
+	err := c.prepareLeaderSelection(next)
+	if err != nil {
+		return flow.ZeroID, fmt.Errorf("could not compute leader selection for next epoch: %w", err)
+	}
```

Comment:
🙏 
I think the implementation returns a sentinel error here 
https://github.com/onflow/flow-go/blob/04b7de4043c5ee80d65c2acff3d51769ccf00c98/state/protocol/errors.go#L16



To document this would be great so external code is able to differentiate between the case:
* view is too far in the future to determine leader selection
* some internal (probably fatal) problem occurred 


#### suggestion

* please add the documentation of the sentinel error return (`ErrNextEpochNotSetup`)  to the `committee` interface
* please implement a unit test which verifies this expected behaviour

Reply:
done in https://github.com/onflow/flow-go/pull/157/commits/8ffaf466ce736b695b83b2beaf567fc649258f3e